{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating TCGA cohorts  (part 1)\n",
    "\n",
    "This notebook will show you how to create a TCGA cohort using the publicly available TCGA BigQuery tables that the [ISB-CGC](http://isb-cgc.org) project has produced based on the open-access [TCGA](http://cancergenome.nih.gov/) data available at the [Data Portal](https://tcga-data.nci.nih.gov/tcga/).  You will need to have access to a Google Cloud Platform (GCP) project in order to use BigQuery.  If you don't already have one, you can sign up for a [free-trial](https://cloud.google.com/free-trial/) or contact [us](mailto://info@isb-cgc.org) and become part of the community evaluation phase of our [Cancer Genomics Cloud pilot](https://cbiit.nci.nih.gov/ncip/nci-cancer-genomics-cloud-pilots).\n",
    "\n",
    "We are not attempting to provide a thorough BigQuery or IPython tutorial here, as a wealth of such information already exists.  Here are some links to some resources that you might find useful: \n",
    "* [BigQuery](https://cloud.google.com/bigquery/what-is-bigquery), \n",
    "* the BigQuery [web UI](https://bigquery.cloud.google.com/) where you can run queries interactively, \n",
    "* [IPython](http://ipython.org/) (now known as [Jupyter](http://jupyter.org/)), and \n",
    "* [Cloud Datalab](https://cloud.google.com/datalab/) the recently announced interactive cloud-based platform that this notebook is being developed on. \n",
    "\n",
    "There are also many tutorials and samples available on github (see, in particular, the [datalab](https://github.com/GoogleCloudPlatform/datalab) repo and the [Google Genomics](  https://github.com/googlegenomics) project).\n",
    "\n",
    "OK then, let's get started!  In order to work with BigQuery, the first thing you need to do is import the bigquery module:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import google.cloud.bigquery as bq\n",
    "client = bq.Client()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next thing you need to know is how to access the specific tables you are interested in.  BigQuery tables are organized into datasets, and datasets are owned by a specific GCP project.  The tables we will be working with in this notebook are in a dataset called **`tcga_201607_beta`**, owned by the **`isb-cgc`** project.  A full table identifier is of the form `<project_id>:<dataset_id>.<table_id>`.  Let's start by getting some basic information about the tables in this dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table ID Annotations                          6322\n",
      "Table ID Biospecimen_data                    23797\n",
      "Table ID Clinical_data                       11160\n",
      "Table ID Copy_Number_segments              2646095\n",
      "Table ID DNA_Methylation_betas          3944304319\n",
      "Table ID DNA_Methylation_chr1            382335670\n",
      "Table ID DNA_Methylation_chr10           197519895\n",
      "Table ID DNA_Methylation_chr11           235823572\n",
      "Table ID DNA_Methylation_chr12           198050739\n",
      "Table ID DNA_Methylation_chr13            97301675\n",
      "Table ID DNA_Methylation_chr14           123239379\n",
      "Table ID DNA_Methylation_chr15           124566185\n",
      "Table ID DNA_Methylation_chr16           179772812\n",
      "Table ID DNA_Methylation_chr17           234003341\n",
      "Table ID DNA_Methylation_chr18            50216619\n",
      "Table ID DNA_Methylation_chr19           211386795\n",
      "Table ID DNA_Methylation_chr2            279668485\n",
      "Table ID DNA_Methylation_chr20            86858120\n",
      "Table ID DNA_Methylation_chr21            35410447\n",
      "Table ID DNA_Methylation_chr22            70676468\n",
      "Table ID DNA_Methylation_chr3            201119616\n",
      "Table ID DNA_Methylation_chr4            159148744\n",
      "Table ID DNA_Methylation_chr5            195864180\n",
      "Table ID DNA_Methylation_chr6            290275524\n",
      "Table ID DNA_Methylation_chr7            240010275\n",
      "Table ID DNA_Methylation_chr8            164810092\n",
      "Table ID DNA_Methylation_chr9             81260723\n",
      "Table ID DNA_Methylation_chrX             98082681\n",
      "Table ID DNA_Methylation_chrY              2330426\n",
      "Table ID Protein_RPPA_data                 1867233\n",
      "Table ID Somatic_Mutation_calls            5356089\n",
      "Table ID mRNA_BCGSC_GA_RPKM                5738048\n",
      "Table ID mRNA_BCGSC_HiSeq_RPKM            38299138\n",
      "Table ID mRNA_UNC_GA_RSEM                 16794358\n",
      "Table ID mRNA_UNC_HiSeq_RSEM             211284521\n",
      "Table ID mRNA_UNC_RSEM                   228078879\n",
      "Table ID miRNA_BCGSC_GA_EXPR_GroupedByMIMAT     943447\n",
      "Table ID miRNA_BCGSC_GA_isoform            6007080\n",
      "Table ID miRNA_BCGSC_GA_mirna              2614260\n",
      "Table ID miRNA_BCGSC_GA_mirna3              738920\n",
      "Table ID miRNA_BCGSC_HiSeq_isoform        46543230\n",
      "Table ID miRNA_BCGSC_HiSeq_mirna          16022160\n",
      "Table ID miRNA_BCGSC_HiSeq_mirna3          6180663\n",
      "Table ID miRNA_Expression                 13226928\n"
     ]
    }
   ],
   "source": [
    "client = bq.Client(project='isb-cgc')\n",
    "dataset = bq.Dataset('isb-cgc.tcga_201607_beta')\n",
    "dataset_ref = client.dataset('tcga_201607_beta')\n",
    "tablestr = \"isb-cgc.tcga_201607_beta.Clinical_data\"\n",
    "\n",
    "tables = list(client.list_tables(dataset))  # API request(s)\n",
    "for table in tables:\n",
    "   table_ref = dataset_ref.table(table.table_id)\n",
    "   table = client.get_table(table_ref)  # API Request\n",
    "   print('Table ID {0:30s} {1:10d}'.format(table.table_id, table.num_rows))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this tutorial, we are going to look at a few different ways that we can use the information in these tables to create cohorts.  Now, you maybe asking what we mean by \"cohort\" and why you might be interested in *creating* one, or maybe what it even means to \"create\" a cohort.  The TCGA dataset includes clinical, biospecimen, and molecular data from over 10,000 cancer patients who agreed to be a part of this landmark research project to build [The Cancer Genome Atlas](http://cancergenome.nih.gov/).  This large dataset was originally organized and studied according to [cancer type](http://cancergenome.nih.gov/cancersselected) but now that this multi-year project is nearing completion, with over 30 types of cancer and over 10,000 tumors analyzed, **you** have the opportunity to look at this dataset from whichever angle most interests you.  Maybe you are particularly interested in early-onset cancers, or gastro-intestinal cancers, or a specific type of genetic mutation.  This is where the idea of a \"cohort\" comes in.  The original TCGA \"cohorts\" were based on cancer type (aka \"study\"), but now you can define a cohort based on virtually any clinical or molecular feature by querying these BigQuery tables.  A cohort is simply a list of samples, using the [TCGA barcode](https://wiki.nci.nih.gov/display/TCGA/TCGA+barcode) system.  Once you have created a cohort you can use it in any number of ways: you could further explore the data available for one cohort, or compare one cohort to another, for example.\n",
    "\n",
    "In the rest of this tutorial, we will create several different cohorts based on different motivating research questions.  We hope that these examples will provide you with a starting point from which you can build, to answer your own research questions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploring the Clinical data table\n",
    "Let's start by looking at the clinical data table.  The TCGA dataset contains a few very basic clinical data elements for almost all patients, and contains additional information for some tumor types only.  For example smoking history information is generally available only for lung cancer patients, and BMI (body mass index) is only available for tumor types where that is a known significant risk factor.  Let's take a look at the clinical data table and see how many different pieces of information are available to us:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{\"type\":\"STRING\",\"name\":\"ParticipantBarcode\"},{\"type\":\"STRING\",\"name\":\"Study\"},{\"type\":\"STRING\",\"name\":\"Project\"},{\"type\":\"STRING\",\"name\":\"ParticipantUUID\"},{\"type\":\"STRING\",\"name\":\"TSSCode\"},{\"type\":\"INTEGER\",\"name\":\"age_at_initial_pathologic_diagnosis\"},{\"type\":\"STRING\",\"name\":\"anatomic_neoplasm_subdivision\"},{\"type\":\"INTEGER\",\"name\":\"batch_number\"},{\"type\":\"STRING\",\"name\":\"bcr\"},{\"type\":\"STRING\",\"name\":\"clinical_M\"},{\"type\":\"STRING\",\"name\":\"clinical_N\"},{\"type\":\"STRING\",\"name\":\"clinical_T\"},{\"type\":\"STRING\",\"name\":\"clinical_stage\"},{\"type\":\"STRING\",\"name\":\"colorectal_cancer\"},{\"type\":\"STRING\",\"name\":\"country\"},{\"type\":\"STRING\",\"name\":\"vital_status\"},{\"type\":\"INTEGER\",\"name\":\"days_to_birth\"},{\"type\":\"INTEGER\",\"name\":\"days_to_death\"},{\"type\":\"INTEGER\",\"name\":\"days_to_last_known_alive\"},{\"type\":\"INTEGER\",\"name\":\"days_to_last_followup\"},{\"type\":\"INTEGER\",\"name\":\"days_to_initial_pathologic_diagnosis\"},{\"type\":\"INTEGER\",\"name\":\"days_to_submitted_specimen_dx\"},{\"type\":\"STRING\",\"name\":\"ethnicity\"},{\"type\":\"STRING\",\"name\":\"gender\"},{\"type\":\"FLOAT\",\"name\":\"gleason_score_combined\"},{\"type\":\"STRING\",\"name\":\"histological_type\"},{\"type\":\"STRING\",\"name\":\"history_of_colon_polyps\"},{\"type\":\"STRING\",\"name\":\"history_of_neoadjuvant_treatment\"},{\"type\":\"STRING\",\"name\":\"hpv_calls\"},{\"type\":\"STRING\",\"name\":\"hpv_status\"},{\"type\":\"STRING\",\"name\":\"icd_10\"},{\"type\":\"STRING\",\"name\":\"icd_o_3_histology\"},{\"type\":\"STRING\",\"name\":\"icd_o_3_site\"},{\"type\":\"STRING\",\"name\":\"lymphatic_invasion\"},{\"type\":\"STRING\",\"name\":\"lymphnodes_examined\"},{\"type\":\"STRING\",\"name\":\"lymphovascular_invasion_present\"},{\"type\":\"STRING\",\"name\":\"menopause_status\"},{\"type\":\"STRING\",\"name\":\"mononucleotide_and_dinucleotide_marker_panel_analysis_status\"},{\"type\":\"STRING\",\"name\":\"neoplasm_histologic_grade\"},{\"type\":\"STRING\",\"name\":\"new_tumor_event_after_initial_treatment\"},{\"type\":\"FLOAT\",\"name\":\"number_of_lymphnodes_examined\"},{\"type\":\"FLOAT\",\"name\":\"number_of_lymphnodes_positive_by_he\"},{\"type\":\"FLOAT\",\"name\":\"number_pack_years_smoked\"},{\"type\":\"INTEGER\",\"name\":\"year_of_initial_pathologic_diagnosis\"},{\"type\":\"STRING\",\"name\":\"pathologic_M\"},{\"type\":\"STRING\",\"name\":\"pathologic_N\"},{\"type\":\"STRING\",\"name\":\"pathologic_T\"},{\"type\":\"STRING\",\"name\":\"pathologic_stage\"},{\"type\":\"STRING\",\"name\":\"person_neoplasm_cancer_status\"},{\"type\":\"STRING\",\"name\":\"pregnancies\"},{\"type\":\"STRING\",\"name\":\"primary_neoplasm_melanoma_dx\"},{\"type\":\"STRING\",\"name\":\"primary_therapy_outcome_success\"},{\"type\":\"FLOAT\",\"name\":\"psa_value\"},{\"type\":\"STRING\",\"name\":\"race\"},{\"type\":\"STRING\",\"name\":\"residual_tumor\"},{\"type\":\"STRING\",\"name\":\"tobacco_smoking_history\"},{\"type\":\"STRING\",\"name\":\"tumor_tissue_site\"},{\"type\":\"STRING\",\"name\":\"tumor_type\"},{\"type\":\"FLOAT\",\"name\":\"weight\"},{\"type\":\"FLOAT\",\"name\":\"height\"},{\"type\":\"FLOAT\",\"name\":\"BMI\"},{\"type\":\"STRING\",\"name\":\"age_began_smoking_in_years\"},{\"type\":\"STRING\",\"name\":\"h_pylori_infection\"},{\"type\":\"STRING\",\"name\":\"other_dx\"},{\"type\":\"STRING\",\"name\":\"other_malignancy_anatomic_site\"},{\"type\":\"STRING\",\"name\":\"other_malignancy_histological_type\"},{\"type\":\"STRING\",\"name\":\"other_malignancy_malignancy_type\"},{\"type\":\"STRING\",\"name\":\"stopped_smoking_year\"},{\"type\":\"STRING\",\"name\":\"venous_invasion\"},{\"type\":\"STRING\",\"name\":\"year_of_tobacco_smoking_onset\"}]\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "bq show --schema isb-cgc:tcga_201607_beta.Clinical_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's a lot of fields!  We can also get at the schema programmatically:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This table has 70 fields. \n",
      "The first few field names and types are: \n",
      "['ParticipantBarcode', 'Study', 'Project', 'ParticipantUUID', 'TSSCode']\n",
      "['STRING', 'STRING', 'STRING', 'STRING', 'STRING']\n"
     ]
    }
   ],
   "source": [
    "table_ref = dataset_ref.table('Clinical_data')\n",
    "table = client.get_table(table_ref)  # API Request\n",
    "#table = bq.Table('isb-cgc:tcga_201607_beta.Clinical_data')\n",
    "#table = client.get_table(table_ref)\n",
    "#if ( table.exists() ):\n",
    "#print(\"Schema is {}\".format(table.schema))\n",
    "fieldNames = map(lambda tsf: tsf.name, table.schema)\n",
    "fieldTypes = map(lambda tsf: tsf.field_type, table.schema)\n",
    "listfieldNames = list(fieldNames)\n",
    "listfieldTypes = list(fieldTypes)\n",
    "\n",
    "print(\"This table has {} fields. \".format( len(listfieldNames) ))\n",
    "print(\"The first few field names and types are: \") \n",
    "print(listfieldNames[:5])\n",
    "print(listfieldTypes[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at these fields and see which ones might be the most \"interesting\", by looking at how many times they are filled-in (not NULL), or how much variation exists in the values.  If we wanted to look at just a single field, \"tobacco_smoking_history\" for example, we could use a very simple query to get a basic summary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tobacco_smoking_history</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>8161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  tobacco_smoking_history     n\n",
       "0                    None  8161\n",
       "1                       1   865\n",
       "2                       4   799\n",
       "3                       2   710\n",
       "4                       3   568\n",
       "5                       5    57"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery \n",
    "\n",
    "SELECT tobacco_smoking_history, COUNT(*) AS n\n",
    "FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "GROUP BY tobacco_smoking_history\n",
    "ORDER BY n DESC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "But if we want to loop over *all* fields and get a sense of which fields might provide us with useful criteria for specifying a cohort, we'll want to automate that.  We'll put a threshold on the minimum number of patients that we expect information for, and the maximum number of unique values (since fields such as the \"ParticipantBarcode\" will be unique for every patient and, although we will need that field later, it's probably not useful for defining a cohort)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " The miRNA_Expression table describes a total of 13226928 patients. \n",
      "Field Names list is 70\n",
      "SELECT ParticipantBarcode FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT Study FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT Project FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT ParticipantUUID FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT TSSCode FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT age_at_initial_pathologic_diagnosis FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT anatomic_neoplasm_subdivision FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT batch_number FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT bcr FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT clinical_M FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT clinical_N FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT clinical_T FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT clinical_stage FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT colorectal_cancer FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT country FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT vital_status FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT days_to_birth FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT days_to_death FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT days_to_last_known_alive FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
      "SELECT days_to_last_followup FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n"
     ]
    }
   ],
   "source": [
    "client = bq.Client()\n",
    "numPatients = table.num_rows\n",
    "print(\" The {} table describes a total of {} patients. \".format(table.table_id, table.num_rows) )\n",
    "\n",
    "# let's set a threshold for the minimum number of values that a field should have,\n",
    "# and also the maximum number of unique values\n",
    "minNumPatients = int(numPatients*0.80)\n",
    "maxNumValues = 50\n",
    "\n",
    "\n",
    "numInteresting = 0\n",
    "iList = []\n",
    "print(\"Field Names list is {}\".format(len(listfieldNames)))\n",
    "for iField in range(len(listfieldNames)):\n",
    "  aField = listfieldNames[iField]\n",
    "  aType = listfieldTypes[iField]\n",
    "  try:\n",
    "    qString = \"SELECT {0} FROM `{1}`\".format(aField,tablestr)\n",
    "    df = client.query(qString).to_dataframe()\n",
    "    summary = df[str(aField)].describe()\n",
    "    if ( aType == \"STRING\" ):\n",
    "      topFrac = float(summary['freq'])/float(summary['count'])\n",
    "      if ( summary['count'] >= minNumPatients ):\n",
    "        if ( summary['unique'] <= maxNumValues and summary['unique'] > 1 ):\n",
    "          if ( topFrac < 0.90 ):\n",
    "            numInteresting += 1\n",
    "            iList += [aField]\n",
    "            print(\"   > {} has {} values with {} unique ({} occurs {} times) \". \\\n",
    "              format(str(aField), summary['count'], summary['unique'], summary['top'], summary['freq']))\n",
    "    else:\n",
    "      if ( summary['count'] >= minNumPatients ):\n",
    "        if ( summary['std'] > 0.1 ):\n",
    "          numInteresting += 1\n",
    "          iList += [aField]\n",
    "          print(\"     > {} has {} values (mean={:0}, sigma={:0} \". \\\n",
    "            format(str(aField), summary['count'], summary['mean'], summary['std']))\n",
    "  except:\n",
    "    pass    \n",
    "\n",
    "print(\" \")\n",
    "print(\" Found {} potentially interesting features: \".format(numInteresting))\n",
    "print(\"   \", iList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"\"\"SELECT ParticipantBarcode FROM `isb-cgc.tcga_201607_beta.Clinical_data` LIMIT 5\"\"\"\n",
    "#results = client.query(query).result()\n",
    "\n",
    "df = client.query(query).to_dataframe()\n",
    "summary = df[str(\"ParticipantBarcode\")].describe()\n",
    "print(summary)\n",
    "#for row in results:\n",
    "#    print(\"{} : \".format(row.ParticipantBarcode))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The above helps us narrow down on which fields are likely to be the most useful, but if you have a specific interest, for example in menopause or HPV status, you can still look at those in more detail very easily: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bigquery\n",
    "SELECT menopause_status, COUNT(*) AS n\n",
    "FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "WHERE menopause_status IS NOT NULL\n",
    "GROUP BY menopause_status\n",
    "ORDER BY n DESC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We might wonder which specific tumor types have menopause information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bigquery\n",
    "SELECT Study, COUNT(*) AS n\n",
    "FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "WHERE menopause_status IS NOT NULL\n",
    "GROUP BY Study\n",
    "ORDER BY n DESC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bigquery\n",
    "SELECT hpv_status, hpv_calls, COUNT(*) AS n\n",
    "FROM `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "WHERE hpv_status IS NOT NULL\n",
    "GROUP BY hpv_status, hpv_calls\n",
    "HAVING n > 20\n",
    "ORDER BY n DESC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TCGA Annotations\n",
    "\n",
    "An additional factor to consider, when creating a cohort is that there may be additional information that might lead one to exclude a particular patient from a cohort.  In certain instances, patients have been redacted or excluded from analyses for reasons such as prior treatment, etc, but since different researchers may have different criteria for using or excluding certain patients or certain samples from their analyses, in many cases the data is still available while at the same time \"annotations\" may have been entered into a searchable [database](https://tcga-data.nci.nih.gov/annotations/).  These annotations have also been uploaded into a BigQuery table and can be used in conjuction with the other BigQuery tables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Early-onset Breast Cancer\n",
    "\n",
    "Now that we have a better idea of what types of information is available in the Clinical data table, let's create a cohort consisting of female breast-cancer patients, diagnosed at the age of 50 or younger."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this next code cell, we define several queries within a **`module`** which allows us to use them both individually and by reference in the final, main query.  \n",
    "+ the first query, called **`select_on_annotations`**, finds all patients in the Annotations table which have either been 'redacted' or had 'unacceptable prior treatment';  \n",
    "+ the second query, **`select_on_clinical`** selects all female breast-cancer patients who were diagnosed at age 50 or younger, while also pulling out a few additional fields that might be of interest;  and\n",
    "+ the final query joins these two together and returns just those patients that meet the clinical-criteria and do **not** meet the exclusion-criteria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ParticipantBarcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-3C-AALI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-4H-AAAK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-5L-AAT0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-A1-A0SH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-A1-A0SJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGA-A1-A0SN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGA-A1-A0SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCGA-A1-A0SQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TCGA-A2-A04P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TCGA-A2-A04Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TCGA-A2-A04R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TCGA-A2-A04U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TCGA-A2-A04V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TCGA-A2-A04W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TCGA-A2-A04X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TCGA-A2-A0CL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TCGA-A2-A0CM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>TCGA-A2-A0CV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>TCGA-A2-A0CZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TCGA-A2-A0D2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>TCGA-A2-A0D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>TCGA-A2-A0D4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>TCGA-A2-A0EX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>TCGA-A2-A0SX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TCGA-A2-A0T3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TCGA-A2-A0T5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TCGA-A2-A0T6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TCGA-A2-A0YE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TCGA-A2-A0YJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TCGA-A2-A0YL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>TCGA-GM-A2DL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>TCGA-GM-A3XG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>TCGA-GM-A3XL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>TCGA-GM-A3XN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>TCGA-HN-A2OB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>TCGA-JL-A3YW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>TCGA-JL-A3YX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>TCGA-LD-A66U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>TCGA-LL-A5YN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>TCGA-LL-A5YO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>TCGA-LL-A5YP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>TCGA-LL-A6FR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>TCGA-LL-A7SZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>TCGA-MS-A51U</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>TCGA-OL-A5D8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>TCGA-OL-A5RV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>TCGA-OL-A5RW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>TCGA-OL-A66I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>TCGA-OL-A66O</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>TCGA-OL-A6VO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>TCGA-OL-A6VQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>TCGA-OL-A6VR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>TCGA-PE-A5DE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>TCGA-PL-A8LX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>TCGA-PL-A8LY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>TCGA-PL-A8LZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>TCGA-S3-A6ZH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>TCGA-S3-AA14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>TCGA-XX-A899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>TCGA-Z7-A8R6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>327 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ParticipantBarcode\n",
       "0         TCGA-3C-AALI\n",
       "1         TCGA-4H-AAAK\n",
       "2         TCGA-5L-AAT0\n",
       "3         TCGA-A1-A0SH\n",
       "4         TCGA-A1-A0SJ\n",
       "5         TCGA-A1-A0SN\n",
       "6         TCGA-A1-A0SP\n",
       "7         TCGA-A1-A0SQ\n",
       "8         TCGA-A2-A04P\n",
       "9         TCGA-A2-A04Q\n",
       "10        TCGA-A2-A04R\n",
       "11        TCGA-A2-A04U\n",
       "12        TCGA-A2-A04V\n",
       "13        TCGA-A2-A04W\n",
       "14        TCGA-A2-A04X\n",
       "15        TCGA-A2-A0CL\n",
       "16        TCGA-A2-A0CM\n",
       "17        TCGA-A2-A0CV\n",
       "18        TCGA-A2-A0CZ\n",
       "19        TCGA-A2-A0D2\n",
       "20        TCGA-A2-A0D3\n",
       "21        TCGA-A2-A0D4\n",
       "22        TCGA-A2-A0EX\n",
       "23        TCGA-A2-A0SX\n",
       "24        TCGA-A2-A0T3\n",
       "25        TCGA-A2-A0T5\n",
       "26        TCGA-A2-A0T6\n",
       "27        TCGA-A2-A0YE\n",
       "28        TCGA-A2-A0YJ\n",
       "29        TCGA-A2-A0YL\n",
       "..                 ...\n",
       "297       TCGA-GM-A2DL\n",
       "298       TCGA-GM-A3XG\n",
       "299       TCGA-GM-A3XL\n",
       "300       TCGA-GM-A3XN\n",
       "301       TCGA-HN-A2OB\n",
       "302       TCGA-JL-A3YW\n",
       "303       TCGA-JL-A3YX\n",
       "304       TCGA-LD-A66U\n",
       "305       TCGA-LL-A5YN\n",
       "306       TCGA-LL-A5YO\n",
       "307       TCGA-LL-A5YP\n",
       "308       TCGA-LL-A6FR\n",
       "309       TCGA-LL-A7SZ\n",
       "310       TCGA-MS-A51U\n",
       "311       TCGA-OL-A5D8\n",
       "312       TCGA-OL-A5RV\n",
       "313       TCGA-OL-A5RW\n",
       "314       TCGA-OL-A66I\n",
       "315       TCGA-OL-A66O\n",
       "316       TCGA-OL-A6VO\n",
       "317       TCGA-OL-A6VQ\n",
       "318       TCGA-OL-A6VR\n",
       "319       TCGA-PE-A5DE\n",
       "320       TCGA-PL-A8LX\n",
       "321       TCGA-PL-A8LY\n",
       "322       TCGA-PL-A8LZ\n",
       "323       TCGA-S3-A6ZH\n",
       "324       TCGA-S3-AA14\n",
       "325       TCGA-XX-A899\n",
       "326       TCGA-Z7-A8R6\n",
       "\n",
       "[327 rows x 1 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery \n",
    "WITH \n",
    "  select_on_annotations AS (\n",
    "    SELECT\n",
    "      ParticipantBarcode,\n",
    "      annotationCategoryName AS categoryName,\n",
    "      annotationClassification AS classificationName\n",
    "    FROM\n",
    "      `isb-cgc.tcga_201607_beta.Annotations`\n",
    "    WHERE\n",
    "      ( itemTypeName=\"Patient\"\n",
    "        AND (annotationCategoryName=\"History of unacceptable prior treatment related to a prior/other malignancy\"\n",
    "          OR annotationClassification=\"Redaction\" ) )\n",
    "    GROUP BY\n",
    "      ParticipantBarcode,\n",
    "      categoryName,\n",
    "      classificationName\n",
    "  ),\n",
    "  select_on_clinical AS (\n",
    "    SELECT\n",
    "      ParticipantBarcode,\n",
    "      vital_status,\n",
    "      days_to_last_known_alive,\n",
    "      ethnicity,\n",
    "      histological_type,\n",
    "      menopause_status,\n",
    "      race\n",
    "    FROM\n",
    "      `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "    WHERE\n",
    "      ( Study=\"BRCA\"\n",
    "        AND age_at_initial_pathologic_diagnosis<=50\n",
    "        AND gender=\"FEMALE\" )\n",
    "  )\n",
    "SELECT\n",
    "  ac.ParticipantBarcode AS ParticipantBarcode\n",
    "FROM  (\n",
    "  SELECT\n",
    "    a.categoryName,\n",
    "    a.classificationName,\n",
    "    c.ParticipantBarcode\n",
    "  FROM select_on_annotations AS a\n",
    "  FULL OUTER JOIN select_on_clinical AS c\n",
    "  ON\n",
    "    a.ParticipantBarcode = c.ParticipantBarcode\n",
    "  WHERE\n",
    "    (a.ParticipantBarcode IS NOT NULL\n",
    "      OR c.ParticipantBarcode IS NOT NULL )\n",
    "  ORDER BY\n",
    "    a.classificationName,\n",
    "    a.categoryName,\n",
    "    a.ParticipantBarcode,\n",
    "    c.ParticipantBarcode ) AS ac\n",
    "WHERE\n",
    "  ( ac.categoryName IS NULL\n",
    "    AND ac.classificationName IS NULL\n",
    "    AND ac.ParticipantBarcode IS NOT NULL )\n",
    "ORDER BY\n",
    "  ac.ParticipantBarcode"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we explicitly call just the first query in the module, and we get a list of 212 patients with one of these disqualifying annotations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#bq.Query(createCohort_and_checkAnnotations.select_on_annotations).results().to_dataframe()\n",
    "\n",
    "select_on_annotations = \"\"\"\n",
    "    SELECT\n",
    "      ParticipantBarcode,\n",
    "      annotationCategoryName AS categoryName,\n",
    "      annotationClassification AS classificationName\n",
    "    FROM\n",
    "      `isb-cgc.tcga_201607_beta.Annotations`\n",
    "    WHERE\n",
    "      ( itemTypeName=\"Patient\"\n",
    "        AND (annotationCategoryName=\"History of unacceptable prior treatment related to a prior/other malignancy\"\n",
    "          OR annotationClassification=\"Redaction\" ) )\n",
    "    GROUP BY\n",
    "      ParticipantBarcode,\n",
    "      categoryName,\n",
    "      classificationName\n",
    "\"\"\"\n",
    "df = client.query(select_on_annotations).to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and here we explicitly call just the second query, resulting in 329 patients:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#bq.Query(createCohort_and_checkAnnotations.select_on_clinical).results().to_dataframe()\n",
    "select_on_clinical = \"\"\"\n",
    "    SELECT\n",
    "      ParticipantBarcode,\n",
    "      vital_status,\n",
    "      days_to_last_known_alive,\n",
    "      ethnicity,\n",
    "      histological_type,\n",
    "      menopause_status,\n",
    "      race\n",
    "    FROM\n",
    "      `isb-cgc.tcga_201607_beta.Clinical_data`\n",
    "    WHERE\n",
    "      ( Study=\"BRCA\"\n",
    "        AND age_at_initial_pathologic_diagnosis<=50\n",
    "        AND gender=\"FEMALE\" )\n",
    "\"\"\"\n",
    "df = client.query(select_on_clinical).to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "and finally we call the main query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#TBD\n",
    "df = bq.Query(createCohort_and_checkAnnotations).results().to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that we didn't need to call each sub-query individually, we could have just called the main query and gotten the same result.  As you can see, two patients that met the clinical select criteria (which returned 329 patients) were excluded from the final result (which returned 327 patients)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before we leave off, here are a few useful tricks for working with BigQuery in Cloud Datalab:\n",
    "+ if you want to see the raw SQL, you can just build the query and then print it out (this might be useful, for example, in debugging a query -- you can copy and paste the SQL directly into the BigQuery Web UI);\n",
    "+ if you want to see how much data and which tables are going to be touched by this data, you can use the \"dry run\" option.  (Notice the \"cacheHit\" flag -- if you have recently done a particular query, you will not be charged to repeat it since it will have been cached.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "q.execute_dry_run()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
